proof/crefine/lib/Corres_C.thy

```diff
@@ -1015,21 +1011,53 @@ lemma ccorres_zipWithM_x_while_genQ:
               (body ;; Basic (\<lambda>s. xf_update (\<lambda>_. xf s + of_nat j) s))))"
   unfolding zipWithM_x_def
   apply (rule ccorres_guard_imp)
-    apply (rule ccorres_rel_imp [OF ccorres_sequence_x_while_genQ[where F=F, OF _ _ _ _ _ xf j]],
-      simp_all add: length_zipWith)
-      apply (simp add: length_zipWith zipWith_nth)
-      apply (rule rl)
-     apply (rule guard)
-    apply (rule bodyi)
-   apply (simp add: zipWith_nth hi[simplified])
-  apply (rule wb)
+    apply (rule ccorres_rel_imp [OF ccorres_sequence_x_while_genQ[where F=F, OF _ _ _ _ _ xf j]];
+           simp)
+        apply (simp add: zipWith_nth)
+        apply (rule rl)
+       apply (rule guard)
+      apply (rule bodyi)
+     apply (simp add: zipWith_nth hi[simplified])
+    apply (rule wb)
+   apply simp+
   done
 
+\<comment> \<open>This context is used to temporarily remove ccorres_weak_cong, so that the following lemmas can
+    be constructed with simplified return relations.
+    We do not use ccorres_all_cong due to it causing unexpected eta-expansion.\<close>
+context
+notes ccorres_weak_cong[cong del]
+begin
```

Comment:
I moved these lemmas so that I only needed the context boilerplate once. Are we ok with this, or would we prefer to keep the original structure with each lemma following the lemma it is constructed from?

Reply:
From my side that's fine, it's different, but no less sane than the original I think

Reply:
I think you can drop "this context is used to" from the doc. Good idea clumping them together. Would they make more sense towards the top of the file rather than at the very bottom though?

Reply:
They need to be after all of the lemmas that they are constructed from, so without much larger changes to the file structure this is the earliest they can go.